optimize/types.go

```diff
@@ -132,13 +124,17 @@ type Problem struct {
 	// must not modify x.
 	Func func(x []float64) float64
 
-	// Grad evaluates the gradient at x and stores the result in-place in grad.
-	// Grad must not modify x.
-	Grad func(grad []float64, x []float64)
+	// Grad evaluates the gradient at x and returns the result. Grad may use
+	// (and return) the provided slice if it is non-nil, or must allocate a new
+	// slice otherwise.
+	Grad func(grad []float64, x []float64) []float64
 
 	// Hess evaluates the Hessian at x and stores the result in-place in hess.
-	// Hess must not modify x.
-	Hess func(hess mat.MutableSymmetric, x []float64)
+	// Hess must not modify x. Hess may use (and return) the provided Symmetric
+	// if it is non-nil, or must allocate a new Symmetric otherwise. Minimize
+	// will 'give back' the returned Symmetric where possible, allowing Hess
```

Comment:
Can we make the statement here stronger than "where possible"?

Reply:
I don't think so, if a Method needs to store Hessian information for some reason, if we guarantee memory re-use then the Method needs to copy.